lib/Infer/InstSynthesis.cpp

```diff
@@ -0,0 +1,1376 @@
+// Copyright 2015 The Souper Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/raw_ostream.h"
+#include "souper/Extractor/KLEEBuilder.h"
+#include "souper/Infer/InstSynthesis.h"
+
+using namespace souper;
+using namespace llvm;
+
+namespace {
+
+static cl::opt<bool> DebugSynthesis("souper-synthesis-debug",
```

Comment:
I think it would make sense to collapse these into a single variable where level 0 means debugging is turned off, etc.